# Task Distribution
| Member | ID | Tasks |
|---|---|---|
|Bhavika| - | - | 
|Alora| - | - |
|Andrea| - | - |

# Import Libraries

In [216]:
%pip install tweepy

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 23.0 -> 23.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [1]:
import tweepy
import sys
import csv
import pandas as pd
import re
import numpy as np

# Scraping Tweets

In [2]:
# Keys for authentication
consumer_key = "ONLQ0vLSAXM2tZDnckominEcK"
consumer_secret = "MpP7XurdGn1wpufX3rrfQgrs4AROnaw9ZiiZzKN6exr3zEDlN6"
access_token = "1274112117738737664-J2DjjoqqzN11CNp8bpnYzVp1dUYKM4"
access_token_secret = "3coYFJVCqHnOmD2m9Qt6zs7j1I24lhlPTU71eUOq4Zqg9"
bearer_token = "AAAAAAAAAAAAAAAAAAAAALJslQEAAAAA7EWdkTwrVDM44LIuewNascHjvoY%3DeXmhQyuXgo8zXJiSTMfqtmPA393HHO2W7nSrk0bK6b67SuJeb0"

In [3]:
# Authenticate
auth = tweepy.OAuth1UserHandler(
    consumer_key, consumer_secret, access_token, access_token_secret)

api = tweepy.API(auth)

In [4]:
def scrape_tweets(query: str, amount: int) -> list:
    """Scrapes a specified number of tweets based on given query and location.

    Method from https://www.sahilfruitwala.com/guide-to-extract-tweets-using-tweepy#how-to-retrieve-specific-number-of-tweets-using-tweepy 
    
    Args:
        query (str): The query
    Returns:
        list (str): List of contents of tweets.
    """
    extracted_tweets = []
    for tweet in tweepy.Cursor(api.search_tweets, query, count=100, tweet_mode="extended", result_type="recent").items(amount):
        extracted_tweets.append(tweet.full_text)
    return extracted_tweets

In [221]:
def write_to_csv(tweets: list, csv_name: str) -> None:
    """Appends the list of tweets to csv.
    
    Largely copied from https://gist.github.com/anku255/0cebd75cce675f2b56de1ef48ec06575.

    """
    print("Writing {0} tweets to {1}".format(len(tweets), csv_name))
    tweets_for_csv = [[tweet.encode("utf-8")] for tweet in tweets]
    with open(csv_name, "a+") as file:
        writer = csv.writer(file, delimiter=",")
        writer.writerows(tweets_for_csv)


The queries for scraping tweets.

In [222]:

queries = [
    ["(metaverse OR meta verse OR #metaverse OR #meta #verse) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(innovation OR #innovation OR innovate OR innovative OR #innovate OR #innovative) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(sustainability OR #sustainability OR sustainable) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(technology footprint OR technology OR #technologyfootprint) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(artificial intelligence OR ai OR #ai OR #artificialintelligence) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(ethics OR #ethics OR ethical OR #ethical) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(inflation OR #inflation) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(cop 28 OR cop28 OR #cop28 OR #cop #28) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(museum of the future OR @museumofthefuture OR #museumofthefuture OR #museum #of #the #future) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(bit coin OR bitcoin OR #bitcoin OR #crypto OR crpyto OR cryptocurrency) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(cyber attack OR #cyberattack OR #cyber OR #cyberattacks OR #cyber #attack) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(robot OR robots OR #robots OR #robot) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(drone OR drones OR #drone OR #drones) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(hacker OR hacking OR #hacker OR #hacking OR #hack) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(chatgpt OR #chatgpt OR #chat #gpt OR chat gpt) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(cybersecurity OR #cybersecurityOR #cybersec OR cybersec OR cyber security OR #cyber #security) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(sustainabletech OR sustainable ai OR sustainable technology OR #sustainabletech OR #sustainableai OR #sustainabletechnology) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(computing OR quantum computing OR #quantum #computing OR #computing) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(automation OR #automation) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(distributed cloud OR #distributedcloud OR #distributed #cloud) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(big data OR #big #data OR bigdata OR #bigdata) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(ar OR #ar OR #augmentedreality OR #augmented #reality) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(data mining OR #data #mining OR #datamining) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(software OR #software) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(tech OR technology OR #tech OR #technology) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(digital OR digital transformation OR #digital #transformation OR #digitaltransformation) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(blockchain OR #blockchain) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(coders_hq OR coders hq OR #coders OR #coders #hq OR @coders_hq) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(nft OR non-fungible token OR non fungible token OR nfts OR #nfts OR #nonfungibletoken OR #nft) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(binance OR @binance OR #binance) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(uae hackathon OR hackathon OR #hackathon) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(internet of things OR IoT OR #iot OR #internet #of #things OR #internetofthings) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(software engineering OR #softwareengineering) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(#coding OR coding) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(@uaeai OR #uaeai Or #uae #ai) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(technology OR tech OT #tech Or #technology) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(nft OR #nft OR #nfts OR nfts) AND (place_country:AE OR uae) lang:en -filter:retweets"]   ,
    ["(UAE codes OR #UAE_codes OR #UAEcodes) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(Microsoft Hololens OR Hololens OR #Hololens) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(GITEX or #gitex) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(information OR #information) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(internet OR #internet) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(computer OR science OR computer science OR #computerscience OR #computer OR #science) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(technological OR #technological) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(engineering OR #engineering) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(computing OR #computing) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(systems OR #systems) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(electronics OR #electronics) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(application OR #application OR app OR #app) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(robotics OR #robotics) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(high tech OR high technology OR #hightech OR #hightechnology OR #high #technology OR #high #tech) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(nanotech OR nano tech OR nanotechnology OR nano technology OR #nanotech OR #nano #tech OR #nano #technology OR #nanotechnology) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(biotech OR bio tech OR biotechnology OR bio technology OR #biotech OR #bio #tech OR #bio #technology OR #biotechnology) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(code OR #code) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(automation OR #automation OR automate OR #automate) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(virtual OR #virtual OR online OR #online) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(technological advancements OR #technological #advancements OR #advancements OR #techadvancements) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(bionics OR #bionics) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(cyberscience OR #cyberscience) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(google OR #google OR @google) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(oracle OR #oracle OR @oracle) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(ibm OR #ibm OR @ibm) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(bionics OR #bionics) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(sap OR #sap) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(global knowledge OR #globalknowledge OR #global #knowledge) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(amazon OR #amazon OR @amazon) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(hewlett-packard OR Hewlett Packard OR HP OR #hewlett-packard OR #Hewlett #Packard OR #HP) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(cisco OR #cisco OR @cisco) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(dell OR #dell OR @dell) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(cisco OR #cisco OR @cisco) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(@WorldGovSummit OR #WGS OR #WGS2023 OR World Goverment Summit) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(masdar OR #masdar OR @masdar) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(renewable energy OR clean energy OR #renewableenergy OR #cleanenergy) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(machine learning OR #machinelearning OR ML OR #ML) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(recommender systems OR #RS OR #recommender #systems OR recommender engine) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(deep learning OR #deeplearning) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(research OR #research) AND (place_country:AE OR uae) lang:en -filter:retweets"],
]

Commented out below to prevent re-running code.

In [223]:
# """Scraping tweets and writing them to dataset.csv"""
# csv_name="dataset.csv"
# total = 0
# for query in queries:
#     tweets = scrape_tweets(query=query, amount=500)
#     total += len(tweets)
#     print("Found {0} tweets related to the query {1}".format(len(tweets), query))
#     write_to_csv(tweets=tweets,csv_name=csv_name)
# print("Done. {0} rows.".format(total))

# Data Cleaning

This section cleans the scraped tweets dataset of duplicate tweets, URLs, @'s and #s.

Note: 
- Do not modify original csv `dataset.csv`. The dataset with no duplicates is stored in `updated_dataset.csv`

In [2]:
# Dropping duplicates
dataset = pd.read_csv("dataset.csv", header=None)
dataset.rename(columns={0: 'Text'}, inplace=True)


In [3]:
dataset

Text
0      b'Joined UAE Minister &amp; President-Designat...
1      b'Foreign Secretary @AmbVMKwatra participated ...
2      b'As the UAE marks National Environment Day, w...
3      b'RT @KhalifaAlgaz: Did you know that the UAE ...
4      b'RT @tawhidChtioui: \xf0\x9f\x9b\xb0\xef\xb8\...
...                                                  ...
21974  b'Send our special gifts to your loved ones!\x...
21975  b'Trainer code: 2735 9457 4031\nfrom \xf0\x9f\...
21976  b"We are looking for ICY SNOW , SANDSTORM .\nI...
21977  b'Make #MONEY from #home through PC #APP \nMak...
21978  b'#Earn Money from #home through PC #APP\nwith...

[21979 rows x 1 columns]

In [4]:
updated_dataset = dataset.drop_duplicates().copy()
updated_dataset


Text
0      b'Joined UAE Minister &amp; President-Designat...
1      b'Foreign Secretary @AmbVMKwatra participated ...
2      b'As the UAE marks National Environment Day, w...
3      b'RT @KhalifaAlgaz: Did you know that the UAE ...
4      b'RT @tawhidChtioui: \xf0\x9f\x9b\xb0\xef\xb8\...
...                                                  ...
21888  b'@m_ut67 @modgovae @Forsan_UAE Dear sir pleas...
21889  b'@UAE_BARQ Dear sir please my help you sir my...
21890  b'@UAE_BARQ Dear sir please my help you sir my...
21915  b"#PhonePe launched a service on Tuesday that ...
21962  b'@WeLoveIndia7 Saudi k log bhikhari h ? \nIn ...

[5878 rows x 1 columns]

### Removing URLs, @'s and hashtags

In [5]:
tweets = updated_dataset['Text'].copy()
updated_dataset['Cleaned Text'] = tweets.str.replace(r'@[^\s]+|#[^\s]+|https?:\/\/\S+|www\.\S+', '',regex=True)
updated_dataset['Cleaned Text'].head()


0    b'Joined UAE Minister &amp; President-Designat...
1    b'Foreign Secretary  participated in a meeting...
2    b'As the UAE marks National Environment Day, w...
3    b'RT  Did you know that the UAE is developing ...
4    b'RT  \xf0\x9f\x9b\xb0\xef\xb8\x8f UAE lunar r...
Name: Cleaned Text, dtype: object

### Removing Duplicates again

In [6]:
updated_dataset = updated_dataset.drop('Text', axis=1)
updated_dataset = updated_dataset.drop_duplicates()
updated_dataset.shape

(5398, 1)

In [7]:
updated_dataset.to_csv('updated-dataset.csv')

In [8]:
updated_dataset = pd.read_csv('updated-dataset.csv', index_col=0)
updated_dataset

Cleaned Text
0      b'Joined UAE Minister &amp; President-Designat...
1      b'Foreign Secretary  participated in a meeting...
2      b'As the UAE marks National Environment Day, w...
3      b'RT  Did you know that the UAE is developing ...
4      b'RT  \xf0\x9f\x9b\xb0\xef\xb8\x8f UAE lunar r...
...                                                  ...
21888  b'   Dear sir please my help you sir my two ki...
21889  b' Dear sir please my help you sir my two kidn...
21890  b' Dear sir please my help you sir my two kidn...
21915  b" launched a service on Tuesday that will all...
21962  b' Saudi k log bhikhari h ? \nIn Middle East, ...

[5398 rows x 1 columns]

# Data Labelling


### TextBlob

For data labelling, we decided to use first use TextBlob. TextBlob calculates the subjectivity and polarity of a text to classify the text as 'Positive', 'Negative' or 'Neutral'

In [9]:
#%pip install TextBlob


In [10]:
from textblob import TextBlob

In [11]:
#Function to get subjectivity
def getSubjectivity(text):
   return TextBlob(text).sentiment.subjectivity
  
 
#Function to get the polarity
def getPolarity(text):
    return TextBlob(text).sentiment.polarity
   
#Create two new columns ‘Subjectivity’ & ‘Polarity’
updated_dataset['TextBlob Subjectivity'] =    updated_dataset['Cleaned Text'].apply(getSubjectivity)
updated_dataset['TextBlob Polarity'] = updated_dataset['Cleaned Text'].apply(getPolarity)
def getAnalysis(score):
  if score < 0:
    return 'Negative'
  elif score == 0:
    return 'Neutral'
  else:
    return 'Positive'

updated_dataset['TextBlob Sentiment'] =  updated_dataset['TextBlob Polarity'].apply(getAnalysis)


In [12]:
updated_dataset['TextBlob Sentiment'].value_counts()

Positive    2643
Neutral     2276
Negative     479
Name: TextBlob Sentiment, dtype: int64

It can be observed that TextBlob classified the majority of tweets to be Positive. To compare, we also decided to use Vader as well to observe if there was a major difference in classification.

### Vader

In [13]:
#%pip install vaderSentiment

In [14]:
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer

sentiment = SentimentIntensityAnalyzer()
def getSentiment(text):
    return sentiment.polarity_scores(text)
    
updated_dataset['Vader Analysis'] = updated_dataset['Cleaned Text'].apply(getSentiment)

def sentimentAnalysis(sentiment_dict):
    if sentiment_dict['compound'] >= 0.05:
        return 'Positive'
    elif sentiment_dict['compound'] <= -0.05:
        return 'Negative'
    else:
        return 'Neutral'

updated_dataset['Vader Sentiment'] = updated_dataset['Vader Analysis'].apply(sentimentAnalysis)

In [15]:
updated_dataset[['Vader Analysis', 'Vader Sentiment']]


Vader Analysis Vader Sentiment
0      {'neg': 0.0, 'neu': 0.638, 'pos': 0.362, 'comp...        Positive
1      {'neg': 0.0, 'neu': 0.816, 'pos': 0.184, 'comp...        Positive
2      {'neg': 0.0, 'neu': 0.766, 'pos': 0.234, 'comp...        Positive
3      {'neg': 0.0, 'neu': 0.874, 'pos': 0.126, 'comp...        Positive
4      {'neg': 0.0, 'neu': 0.819, 'pos': 0.181, 'comp...        Positive
...                                                  ...             ...
21888  {'neg': 0.073, 'neu': 0.726, 'pos': 0.202, 'co...        Positive
21889  {'neg': 0.073, 'neu': 0.726, 'pos': 0.202, 'co...        Positive
21890  {'neg': 0.074, 'neu': 0.719, 'pos': 0.206, 'co...        Positive
21915  {'neg': 0.043, 'neu': 0.774, 'pos': 0.183, 'co...        Positive
21962  {'neg': 0.196, 'neu': 0.777, 'pos': 0.028, 'co...        Negative

[5398 rows x 2 columns]

In [16]:
updated_dataset['Vader Sentiment'].value_counts()

Positive    3335
Neutral     1691
Negative     372
Name: Vader Sentiment, dtype: int64

## Analysis between Vader and TextBlob


| Data Labelling Method      | Positive Label | Negative Label | Neutral Label |
| ----------- | ----------- | ----------- | ----------- |
| TextBlob      | 2643   (49%)   |479  (8.9%)| 2276 (42.1%)|
| Vader   | 3335  (61.8%)      | 372 (6.9%) | 1691 (31.3%)|

From the summary table, we can observe that the negatively labelled tweets stay within a 2% difference of each other, regardless of the data labelling method. However, the positive and neutral label differ by 10%.

In [17]:
# Creating a csv of tweets with conflicting sentiments
unclearTweets = updated_dataset.loc[(updated_dataset['Vader Sentiment'] != updated_dataset['TextBlob Sentiment'])]
unclearTweets.to_csv('ConflictingLabels.csv')

After manually reviewing the differences in labels, it seems that the TextBlob sentiment labelling is more accurate.

# Data Preprocessing

**approach → first data cleaning + tokenization + lemmatization**

## Additional Data Cleaning

### Removing Non-English Characters

In [18]:
def nonenglish(tweet):
    tweet = re.sub(r'\\x[a-zA-Z0-9]+', '', tweet) # tweets in different language changed to hex code \xhh
    tweet = re.sub(r'\\n', '', tweet) # removing new line character as well
    return tweet

updated_dataset['Cleaned Text'] = updated_dataset['Cleaned Text'].apply(lambda tweet: nonenglish(tweet))
updated_dataset['Cleaned Text'].head()


0    b'Joined UAE Minister &amp; President-Designat...
1    b'Foreign Secretary  participated in a meeting...
2    b'As the UAE marks National Environment Day, w...
3    b'RT  Did you know that the UAE is developing ...
4    b'RT   UAE lunar rover will test 1st artificia...
Name: Cleaned Text, dtype: object

### Removing Stopwords

In [19]:
from sklearn.feature_extraction import text

# contains 318 stopwords, including but not limited to articles and prepositions
stopwords = text.ENGLISH_STOP_WORDS

updated_dataset['Cleaned Text'] = updated_dataset['Cleaned Text'].apply(lambda tweet: ' '.join([text for text in tweet.split(' ') if text not in stopwords]))
updated_dataset['Cleaned Text'].head()

0    b'Joined UAE Minister &amp; President-Designat...
1    b'Foreign Secretary  participated meeting foca...
2    b'As UAE marks National Environment Day, remai...
3    b'RT  Did know UAE developing Arabic ChatGPT u...
4    b'RT   UAE lunar rover test 1st artificial int...
Name: Cleaned Text, dtype: object

### Removing Symbols and Punctuations

In [20]:
import string

punctuations = string.punctuation
print(punctuations)

updated_dataset['Cleaned Text'] = updated_dataset['Cleaned Text'].apply(lambda tweet: tweet.translate({ord(punc): None for punc in punctuations}))
updated_dataset['Cleaned Text'].head()

!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~


0    bJoined UAE Minister amp PresidentDesignate  H...
1    bForeign Secretary  participated meeting focal...
2    bAs UAE marks National Environment Day remain ...
3    bRT  Did know UAE developing Arabic ChatGPT us...
4    bRT   UAE lunar rover test 1st artificial inte...
Name: Cleaned Text, dtype: object

### Removing Numbers

In [21]:
updated_dataset['Cleaned Text'] = updated_dataset['Cleaned Text'].apply(lambda tweet: re.sub('[0-9]+', '', tweet))
updated_dataset['Cleaned Text'].head()

0    bJoined UAE Minister amp PresidentDesignate  H...
1    bForeign Secretary  participated meeting focal...
2    bAs UAE marks National Environment Day remain ...
3    bRT  Did know UAE developing Arabic ChatGPT us...
4    bRT   UAE lunar rover test st artificial intel...
Name: Cleaned Text, dtype: object

### Removing 'b' and 'RT'

In [22]:
updated_dataset['Cleaned Text'] = updated_dataset['Cleaned Text'].apply(lambda tweet: re.sub('bRT|b', '', tweet))
updated_dataset['Cleaned Text'].head()

0    Joined UAE Minister amp PresidentDesignate  HE...
1    Foreign Secretary  participated meeting focal ...
2    As UAE marks National Environment Day remain c...
3      Did know UAE developing Araic ChatGPT using ...
4       UAE lunar rover test st artificial intellig...
Name: Cleaned Text, dtype: object

### Tokenization

In [23]:
# as data has been completely cleaned, we can simply tokenize by converting strings to lists
updated_dataset['Tokenized Text'] = updated_dataset['Cleaned Text'].apply(lambda tweet: tweet.split())
updated_dataset['Tokenized Text'].head()

0    [Joined, UAE, Minister, amp, PresidentDesignat...
1    [Foreign, Secretary, participated, meeting, fo...
2    [As, UAE, marks, National, Environment, Day, r...
3    [Did, know, UAE, developing, Araic, ChatGPT, u...
4    [UAE, lunar, rover, test, st, artificial, inte...
Name: Tokenized Text, dtype: object

### Lemmatization

In [24]:
import nltk
lemmatizing = nltk.stem.WordNetLemmatizer()

updated_dataset['Lemmatized Text'] = updated_dataset['Tokenized Text'].apply(lambda tweet: [lemmatizing.lemmatize(text).lower() for text in tweet])
updated_dataset['Lemmatized Text'].head()

0    [joined, uae, minister, amp, presidentdesignat...
1    [foreign, secretary, participated, meeting, fo...
2    [as, uae, mark, national, environment, day, re...
3    [did, know, uae, developing, araic, chatgpt, u...
4    [uae, lunar, rover, test, st, artificial, inte...
Name: Lemmatized Text, dtype: object

In [25]:
# Encode TextBlob Sentiments into numbers
updated_dataset['Encoded_TextBlob_Sentiment'] = updated_dataset['TextBlob Sentiment'].apply(lambda x: 1 if x=='Positive' else -1 if x=='Negative' else 0)

In [26]:
updated_dataset.reset_index(inplace=True)

# Text Representation

This section experiments with different document representations, namely:
- Bag of Words
- N-grams
- TF-IDF
- CBOW
- Skip-gram
- Pre-trained Word2Vec model by Google


Each of the representations are testing using Naive Bayes, and the high performance models are used in the next stage of the pipeline.

Note: Arbitrarily using lemmatization and textblob sentiments

In [27]:
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import classification_report
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.pipeline import Pipeline


In [28]:
X = updated_dataset['Lemmatized Text']
y = updated_dataset['Encoded_TextBlob_Sentiment']
X_train, X_test, y_train, y_test = train_test_split(
    updated_dataset['Lemmatized Text'], 
    updated_dataset['Encoded_TextBlob_Sentiment'],
    test_size=0.2, 
    random_state=42, 
    stratify=updated_dataset['Encoded_TextBlob_Sentiment']
)

### Bag Of Words

In [29]:


bow = CountVectorizer(
    ngram_range= (1,1), # unigram
    preprocessor= lambda x: x, # override preprocessing
    tokenizer= lambda x: x, # override tokenization
)

nb_bow = Pipeline([
    ('bow', bow),
    ('naive_bayes', MultinomialNB())
])



In [30]:
nb_bow.fit(X_train, y_train)
y_pred_bow = nb_bow.predict(X_test)
print(classification_report(y_test, y_pred_bow))

              precision    recall  f1-score   support

          -1       0.71      0.28      0.40        96
           0       0.85      0.58      0.69       455
           1       0.67      0.92      0.77       529

    accuracy                           0.72      1080
   macro avg       0.74      0.60      0.62      1080
weighted avg       0.75      0.72      0.71      1080



### N-Grams
How to find best n-gram number?

In [31]:
ngram = CountVectorizer(
    ngram_range= (1,3), # trigram
    preprocessor= lambda x: x, # override preprocessing
    tokenizer= lambda x: x, # override tokenization
)

ngram_nb = Pipeline([
    ('ngram', ngram),
    ('naive_bayes', MultinomialNB())
])

ngram_nb.fit(X_train, y_train)
ngram_nb_y_pred = ngram_nb.predict(X_test)
print(classification_report(y_test, ngram_nb_y_pred))

              precision    recall  f1-score   support

          -1       0.77      0.35      0.49        96
           0       0.87      0.62      0.72       455
           1       0.69      0.93      0.79       529

    accuracy                           0.75      1080
   macro avg       0.78      0.63      0.67      1080
weighted avg       0.77      0.75      0.74      1080



### TF-IDF

In [32]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf = TfidfVectorizer(
    preprocessor= lambda x: x, # override preprocessing
    tokenizer= lambda x: x, # override tokenization
)

tfidf_nb = Pipeline([
    ('tfidf', tfidf),
    ('naive_bayes', MultinomialNB())
])

tfidf_nb.fit(X_train, y_train)
tfidf_nb_y_pred = tfidf_nb.predict(X_test)
print(classification_report(y_test, tfidf_nb_y_pred))


              precision    recall  f1-score   support

          -1       1.00      0.09      0.17        96
           0       0.88      0.56      0.68       455
           1       0.65      0.95      0.77       529

    accuracy                           0.71      1080
   macro avg       0.84      0.53      0.54      1080
weighted avg       0.77      0.71      0.68      1080



## Word2Vec

In [33]:
from gensim.models import Word2Vec
import gensim

#%pip install gensim

Note: you may need to restart the kernel to use updated packages.


### CBOW


Possible parameters to tweak is window, cbow

In [34]:
cbow = Word2Vec(
    sentences=X,
    window=10, # how many context words before and after target word to consider
    min_count=2,
    workers=4, # number of threads
    sg=0, # 0 to use cbow, 1 for skipgram
    cbow_mean=0, # 0 uses sum of context word vectors, 1 uses mean
)
cbow.train(X, total_examples=cbow.corpus_count,epochs=cbow.epochs)
# After training, KeyedVector object is used
cbow_model = cbow.wv

In [35]:
from sklearn.preprocessing import MinMaxScaler

def normalize_vectors(input_vectors):
    """Normalizes vectors to be in range [0,1)
    
    Args:
        input_vectors (np.ndarray): List of vectors, each vector is np.ndarray
    """

    return MinMaxScaler().fit(input_vectors).transform(input_vectors)

In [36]:
def vectorize_sentence(sentence, wv_model):
    """Convert a sentence to a vector by summing word vectors in the sentence.

    Another option could be to average word vectors rather than just sum.
    
    Arg: 
        sentence (list): Tokenized sentence
    Returns:
        (np.ndarray): Vector representation of sentence
    """
    vector_size = wv_model.vector_size
    wv_res = np.zeros(vector_size)
    count = 0
    for w in sentence:
        if w in wv_model:
            # count += 1
            wv_res += wv_model[w]
    # wv_res = wv_res/count
    return wv_res

In [37]:
# Get the embeddings of train and test set
X_train_cbow = normalize_vectors(np.array([vectorize_sentence(x,cbow_model) for x in X_train]))
X_test_cbow = normalize_vectors(np.array([vectorize_sentence(x,cbow_model) for x in X_test]))

#test NB
nb_cbow = MultinomialNB()
nb_cbow.fit(X_train_cbow, y_train)
y_pred_cbow = nb_cbow.predict(X_test_cbow)
print(classification_report(y_test, y_pred_cbow, zero_division=0))

              precision    recall  f1-score   support

          -1       0.00      0.00      0.00        96
           0       0.62      0.47      0.53       455
           1       0.58      0.81      0.67       529

    accuracy                           0.59      1080
   macro avg       0.40      0.42      0.40      1080
weighted avg       0.54      0.59      0.55      1080



### Skip Gram

In [38]:
skipgram = Word2Vec(
    sentences=X,
    window=10,
    min_count=2,
    workers=4,
    sg=1, # 0 to use cbow, 1 for skipgram
)
# Train the model
skipgram.train(X, total_examples=skipgram.corpus_count,epochs=skipgram.epochs)

# After training, KeyedVector object is used
skipgram_model = skipgram.wv

In [39]:
# Get the embeddings of train and test set
X_train_sg = normalize_vectors(np.array([vectorize_sentence(x,skipgram_model) for x in X_train]))
X_test_sg = normalize_vectors(np.array([vectorize_sentence(x,skipgram_model) for x in X_test]))
#test on NB
nb_sg = MultinomialNB()
nb_sg.fit(X_train_sg, y_train)
y_pred_sg = nb_sg.predict(X_test_sg)
print(classification_report(y_test, y_pred_sg, zero_division=0))

              precision    recall  f1-score   support

          -1       0.00      0.00      0.00        96
           0       0.66      0.46      0.55       455
           1       0.58      0.83      0.68       529

    accuracy                           0.60      1080
   macro avg       0.41      0.43      0.41      1080
weighted avg       0.56      0.60      0.56      1080



### Using Pre-trained Word2Vec Model
Using pre-trained word2vec model from Google, which has 3 million words. Trained on 100 billion words from the google news dataset. Unsure if it uses CBOW or SkipGram or both

From https://thinkingneuron.com/how-to-classify-text-using-word2vec/
Google Model can be downloaded from https://thinkingneuron.com/how-to-classify-text-using-word2vec/#:~:text=download%20link%3A%20https%3A%2F%2Fdrive.google.com%2Ffile%2Fd%2F0b7xkcwpi5kdynlnuttlss21pqmm%2Fedit%3Fusp%3Dsharing

In [40]:
# load the word vectors from Google Model (1GB)
# Will take time because 3 mil words. 
# Download model from https://thinkingneuron.com/how-to-classify-text-using-word2vec/#:~:text=download%20link%3A%20https%3A%2F%2Fdrive.google.com%2Ffile%2Fd%2F0b7xkcwpi5kdynlnuttlss21pqmm%2Fedit%3Fusp%3Dsharing
GoogleModel = gensim.models.KeyedVectors.load_word2vec_format('GoogleNews-vectors-negative300.bin.gz', binary=True,)

FileNotFoundError: [Errno 2] No such file or directory: 'GoogleNews-vectors-negative300.bin.gz'

In [ ]:
def vectorize_sentence_2(sentences, wv_model):
    """Convert a list of sentences to a vectors.

    By summing word vectors in the sentence.
    Another option could be to average word vectors rather than just sum.
    
    Arg: 
        sentence (pd.Series): Series consisting of list of tokenized texts 
    Returns:
        (np.ndarray): Vector representation of sentences
    """
    vector_size = wv_model.vector_size
    count = 0
    vector_sentences = []

    for i in range(len(sentences)): 
        sentence = np.zeros(vector_size)
        for word in sentences[i]:
            count+=1
            if word in wv_model:
                sentence += wv_model[word]
        # sentence = sentence/count
        vector_sentences.append(sentence)
    vector_sentences_df = pd.DataFrame(vector_sentences)
    
    return vector_sentences_df

Convert text data to W2V vectors

In [ ]:
w2v_data = vectorize_sentence_2(X, GoogleModel)
w2v_data.reset_index(inplace=True, drop=True)
w2v_data['sentiment']=y

In [ ]:
w2v_data

0         1         2         3         4         5         6  \
0    -1.132812  2.405243  0.206055  0.998474  0.067627 -0.279907 -1.275879   
1    -2.202881  1.526184  1.248779  0.022171 -0.643616 -0.535217  1.304497   
2    -1.514006  2.221619  1.219940  2.058777 -1.362762 -0.093506  2.393387   
3     0.004150  0.134628  0.555664  0.677246 -0.599525  0.236725  0.374146   
4    -1.098145  0.179138 -0.097778  1.079620 -0.130381 -0.615356  1.475098   
...        ...       ...       ...       ...       ...       ...       ...   
5393 -0.848206  1.854889  0.927032  2.490479 -1.490753  0.884399  0.667587   
5394 -0.848206  1.854889  0.927032  2.490479 -1.490753  0.884399  0.667587   
5395 -0.848206  1.854889  0.927032  2.490479 -1.490753  0.884399  0.667587   
5396 -0.170532  0.555901  0.524551  1.156067 -0.659790  0.719757 -0.405373   
5397 -0.742981  1.656494  1.001097  3.151978 -0.668518 -0.188110 -1.311646   

             7         8         9  ...       291       292       293  \
0    -4.573059  2.728638  1.771240  ...  2.500275 -2.007019  1.403908   
1    -1.866699  1.610382  0.834839  ... -0.429688  0.471130  0.044685   
2    -2.217712  2.003166  1.495361  ...  1.928314 -2.390686  0.371109   
3    -1.129639  0.325989  0.212769  ...  0.318176 -1.261230  0.185188   
4    -1.155151  0.913147  0.649658  ...  2.099121 -0.911377  1.061813   
...        ...       ...       ...  ...       ...       ...       ...   
5393  0.053528  2.230713  1.627228  ... -0.471985 -3.151123 -0.589844   
5394  0.053528  2.230713  1.627228  ... -0.471985 -3.151123 -0.589844   
5395  0.053528  2.230713  1.627228  ... -0.471985 -3.151123 -0.589844   
5396 -0.833130  0.974854  0.811157  ...  1.457672  0.223541  0.326981   
5397 -2.395264  0.448215  2.064152  ...  2.906128 -0.707062  1.330200   

           294       295       296       297       298       299  sentiment  
0    -3.793671 -1.133545 -1.440094 -0.750305  0.297852  4.358887          1  
1    -0.072174  0.426514  1.126831 -0.328308  0.306915  1.507935         -1  
2    -0.576813  0.328003 -0.169029 -0.052216  1.480957 -0.169128          1  
3    -0.322174  0.256348  0.832886  0.004272  0.159027  0.760498          0  
4    -0.402344  0.636719 -1.323242 -1.282776  0.060547 -0.279785         -1  
...        ...       ...       ...       ...       ...       ...        ...  
5393  1.288662  1.495911 -2.537903 -1.745911  0.368065 -0.410278          0  
5394  1.288662  1.495911 -2.537903 -1.745911  0.368065 -0.410278          0  
5395  1.288662  1.495911 -2.537903 -1.745911  0.368065 -0.410278          0  
5396  0.550018 -0.178589  0.676849 -0.140881  0.202820 -0.080078         -1  
5397 -1.054474 -1.403137 -2.133179 -1.052002  0.674034  0.739746          1  

[5398 rows x 301 columns]

In [ ]:
# Get train and test values
X_w2v = w2v_data.iloc[:,:-1].values
y_w2v = w2v_data.iloc[:,-1].values

X_w2v_scaled = normalize_vectors(X_w2v)

X_train_w2v, X_test_w2v, y_train_w2v, y_test_w2v = train_test_split(
    X_w2v_scaled, y_w2v, test_size=0.2, random_state=42,
    stratify= y)

In [ ]:
nb = MultinomialNB()
nb.fit(X_train_w2v, y_train_w2v)
y_pred_w2v = nb.predict(X_test_w2v)
print(classification_report(y_test_w2v, y_pred_w2v, zero_division=0))



              precision    recall  f1-score   support

          -1       0.00      0.00      0.00        96
           0       0.67      0.65      0.66       455
           1       0.65      0.78      0.71       529

    accuracy                           0.65      1080
   macro avg       0.44      0.48      0.45      1080
weighted avg       0.60      0.65      0.62      1080



### Comparison of Text Representation Models
The different models are compared based on their weighted precision, recall, and f1 score.
Values that are closer to one are optimal.

|Model|Precision|Recall|F-1|
|---|---|---|---|
|Bag Of Words (Unigram)|0.75|0.73|0.71|
|Trigrams|0.77|0.75|0.73|
|TF-IDF|0.77|0.71|0.68|
|CBOW (Trained on scraped tweets corpus)|0.58|0.57|0.5|
|Skip-gram (Trained on scraped tweets corpus)|0.6|0.60|0.54|
|Word2Vec model (Trained on Google News Corpus)|0.59|0.65|0.62|


It was expected that the more sophisticated models - those which take into account the context of words - would perform better. However, based on the results the more primitive approaches performed better. This could be due to our small dataset.

The tri-gram model was the best performer in terms of F-1 score, which is the harmonic mean of the precision and recall values.




# Classification Models

### 1. Logistic Regression with Trigrams

In [ ]:
from sklearn.linear_model import LogisticRegression

ngram = CountVectorizer(
    ngram_range= (1, 3), # trigram
    preprocessor= lambda x: x, # override preprocessing
    tokenizer= lambda x: x, # override tokenization
)

ngram_lr = Pipeline([
    ('ngram', ngram),
    ('logistic_regression', LogisticRegression())
])

ngram_lr.fit(X_train, y_train)
ngram_lr_y_pred = ngram_lr.predict(X_test)
print(classification_report(y_test, ngram_lr_y_pred))

              precision    recall  f1-score   support

          -1       0.80      0.38      0.51        96
           0       0.77      0.86      0.81       455
           1       0.82      0.81      0.82       529

    accuracy                           0.79      1080
   macro avg       0.80      0.68      0.71      1080
weighted avg       0.80      0.79      0.79      1080



### 2. FastText with Trigrams

FastText is an open-source library created by Facebook AI Research lab. Setup instructions are here: https://fasttext.cc/docs/en/supervised-tutorial.html#installing-fasttext

In [41]:
#pip install fasttext

In [42]:
import fasttext
help(fasttext.FastText)


Help on module fasttext.FastText in fasttext:

NAME
    fasttext.FastText

DESCRIPTION
    # Copyright (c) 2017-present, Facebook, Inc.
    # All rights reserved.
    #
    # This source code is licensed under the MIT license found in the
    # LICENSE file in the root directory of this source tree.

FUNCTIONS
    cbow(*kargs, **kwargs)
    
    eprint(*args, **kwargs)
    
    load_model(path)
        Load a model given a filepath and return a model object.
    
    read_args(arg_list, arg_dict, arg_names, default_values)
    
    skipgram(*kargs, **kwargs)
    
    supervised(*kargs, **kwargs)
    
    tokenize(text)
        Given a string of text, tokenize it and return a list of tokens
    
    train_supervised(*kargs, **kwargs)
        Train a supervised model and return a model object.
        
        input must be a filepath. The input text does not need to be tokenized
        as per the tokenize function, but it must be preprocessed and encoded
        as UTF-8. You might wan

Reformatting the dataset as FastText expects each row of data to be in the form:
__label__sentiment tweet 

In [51]:
ftDf= updated_dataset[['TextBlob Sentiment', 'Cleaned Text']].copy()
ftDf.iloc[:, 0] = ftDf.iloc[:, 0].apply(lambda x: '__label__' + x)
ftDf.shape

(5398, 2)

In [55]:
#Splitting into test and train
training = ftDf.sample(frac = 0.7)
testing = ftDf.drop(training.index)

training.to_csv('train.txt', 
                index = False, 
                sep = ' ',
                header = None, 
                quoting = csv.QUOTE_NONE, 
                quotechar = "", 
                escapechar = " ")

testing.to_csv('test.txt', 
                index = False, 
                sep = ' ',
                header = None, 
                quoting = csv.QUOTE_NONE, 
                quotechar = "", 
                escapechar = " ")

In [56]:
ftModel = fasttext.train_supervised('train.txt', wordNgrams = 3)

In [59]:
ftModel.test('test.txt')

(1619, 0.44224830142063004, 0.44224830142063004)

To evaluate further, we can test it on unseen samples.

In [82]:
ftModel.predict('I love the UAE for its innovation with artificial intelligence.')

(('__label__Neutral',), array([0.46955368]))

On the test set, fastText model had a precision score of 0.44 ad a recall score of 0.44. It had also misclassified an unseen sample. However, FastText model parameters can be adjusted, so we will try adjusting the epoch and learning rate.

In [81]:
ftModel2 = fasttext.train_supervised('train.txt', wordNgrams = 3, epoch = 6,lr=0.25)
ftModel2.test('test.txt')

(1619, 0.7794935145151328, 0.7794935145151328)

In [83]:
ftModel2.predict('I love the UAE for its innovation with artificial intelligence.')

(('__label__Positive',), array([0.68560129]))

As shown above, increasing the epoch and learning rate corrected a misclassification that happened with the previous model.
